[DataLoader] Implement FilterIterDataPipe

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51880 [DataLoader] Rename files of functional datapipes
* #51879 [DataLoader] Rename Callable to Map IterDataPipe
* **#51783 [DataLoader] Implement FilterIterDataPipe**

Implement Filter to skip element based on customized function which returns boolean value.

Differential Revision: [D26277688](https://our.internmc.facebook.com/intern/diff/D26277688)

ejguan committed:
[WIP][DataLoader] Implement FilterIterDataPipe

[ghstack-poisoned]

glaringlee commented on torch/utils/data/datapipes/iter/sampler.py:
```diff
@@ -1,9 +1,46 @@
+                 ) -> None:
+        super().__init__()
+        self.datapipe = datapipe
+        if filter_fn.__name__ == '<lambda>':
```
nit: It seems this is a common check for all datapipes that have function inputs. Maybe we can put a generic function somewhere, so we don't need to dup the code for this check.

ejguan replied:
Great point! Will update it using Callable as Base class.

glaringlee reviewed (approved):
LGTM. nit comment, not blocking issue.

ejguan committed:
Update on "[DataLoader] Implement FilterIterDataPipe"


Implement Filter to skip element based on customized function which returns boolean value.

Differential Revision: [D26277688](https://our.internmc.facebook.com/intern/diff/D26277688)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Implement FilterIterDataPipe"


Implement Filter to skip element based on customized function which returns boolean value.

Differential Revision: [D26277688](https://our.internmc.facebook.com/intern/diff/D26277688)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Implement FilterIterDataPipe"


Implement Filter to skip element based on customized function which returns boolean value.

Differential Revision: [D26277688](https://our.internmc.facebook.com/intern/diff/D26277688)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Implement FilterIterDataPipe"


Implement Filter to skip element based on customized function which returns boolean value.

Differential Revision: [D26277688](https://our.internmc.facebook.com/intern/diff/D26277688)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Implement FilterIterDataPipe"


Implement Filter to skip element based on customized function which returns boolean value.

Differential Revision: [D26277688](https://our.internmc.facebook.com/intern/diff/D26277688)

[ghstack-poisoned]

VitalyFedyunin commented:
My only concern is the fact that ```FilterIterDataPipe(source_dp, 5, filter_fn = aaaa)``` are confusing and we can do it as ```FilterIterDataPipe(source_dp, filter_fn = aaaa, fn_args = (5, ), fn_kwargs = ... )```

Alternative option is to depend on `import dill` and allow lambdas with bound vars (need research).

ejguan commented:
> My only concern is the fact that `FilterIterDataPipe(source_dp, 5, filter_fn = aaaa)` is confusing and we can do it as `FilterIterDataPipe(source_dp, filter_fn = aaaa, fn_args = (5, ), fn_kwargs = ... )`

You are right, it's confusing. I will change it in #51879  for the behavior of MapIterDataPipe and derived DataPipe

> Alternative option is to depend on `import dill` and allow lambdas with bound vars (need research).

Yes, I did some experiments to use `dill` for lambdas and it works. As we can define how to save it in the `torch.save`, it's doable. But, I think it's better to keep the warning for now as it explicitly shows it doesn't support Pickle. When we implement the `torch.save` for DataPipes/DataLoader, we can modify the warning.
